File: Algorithms/Python/889 Construct Binary Tree from Preorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def constructFromPrePost(self, pre: List[int], post: List[int]) -> TreeNode:
        if len(pre) == 0:
            return None
        
        root = TreeNode(pre[0])
        if len(pre) > 1:
            # pre[1] is the root node of left
            n = post.index(pre[1]) + 1
            root.left = self.constructFromPrePost(pre[1: n + 1], post[: n])
            root.right = self.constructFromPrePost(pre[n + 1:], post[n:])
        return root

File: Algorithms/CPP/889 Construct Binary Tree from Preorder and Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructFromPrePost(vector<int>& pre, vector<int>& post) {
        if (pre.size() == 0) {
            return NULL;
        }

        TreeNode* root = new TreeNode(pre[0]);
        if (pre.size() > 1) {
            int n = std::distance(post.begin(), std::find(post.begin(), post.end(), pre[1])) + 1;
            vector<int> leftPre(pre.begin() + 1, pre.begin() + n + 1);
            vector<int> leftPost(post.begin(), post.begin() + n);
            vector<int> rightPre(pre.begin() + n + 1, pre.end());
            vector<int> rightPost(post.begin() + n, post.end() - 1);
            root->left = constructFromPrePost(leftPre, leftPost);
            root->right = constructFromPrePost(rightPre, rightPost);
        }
        return root;
    }
};
